File: ios/MitiMaiti/ViewModels/InboxViewModel.swift
import SwiftUI

@MainActor
class InboxViewModel: ObservableObject {
    @Published var likes: [LikedYouCard] = []
    @Published var matches: [Match] = []
    @Published var isLoading = false
    @Published var error: String?

    private let api = APIService.shared

    var totalLikes: Int { likes.count }
    var totalMatches: Int { matches.count }
    var unreadMessages: Int { matches.reduce(0) { $0 + $1.unreadCount } }

    private var previousLikeIds: Set<String> = []

    init() {
        // Real-time: the backend emits new_match / match_update over the
        // socket. Refresh the inbox and surface an in-app notification even
        // when the user isn't on the Discover screen. (These streams are
        // single-consumer; InboxViewModel is their only subscriber.)
        Task { @MainActor [weak self] in
            for await payload in SocketChat.shared.newMatches.stream {
                guard let self else { break }
                let name = (payload["displayName"] as? String) ?? "Someone"
                NotificationManager.shared.addNotification(
                    type: .match,
                    title: "New Match!",
                    body: "You and \(name) matched! Say hi before the timer runs out.",
                    actionData: payload["userId"] as? String
                )
                self.loadInbox()
            }
        }
        Task { @MainActor [weak self] in
            for await payload in SocketChat.shared.matchUpdates.stream {
                guard let self else { break }
                guard let matchId = payload["matchId"] as? String,
                      (payload["status"] as? String) == "active" else { continue }
                self.activateMatch(id: matchId)
            }
        }
    }

    func loadInbox() {
        guard !isLoading else { return }
        isLoading = true
        error = nil

        Task {
            do {
                let result = try await api.fetchInbox()
                likes = result.likes
                matches = result.matches
                isLoading = false

                // Create notifications for new likes we haven't seen before
                let currentLikeIds = Set(likes.map(\.id))
                let newLikeIds = currentLikeIds.subtracting(previousLikeIds)
                if !previousLikeIds.isEmpty {
                    for like in likes where newLikeIds.contains(like.id) {
                        NotificationManager.shared.addNotification(
                            type: .like,
                            title: "\(like.user.displayName) liked your profile",
                            body: "Check Liked You to see who!",
                            actionData: like.user.id
                        )
                    }
                }
                previousLikeIds = currentLikeIds

                // Schedule expiry reminders for matches with expiration
                for match in matches {
                    if let expiresAt = match.expiresAt, expiresAt > Date() {
                        NotificationManager.shared.scheduleExpiryReminder(
                            matchName: match.otherUser.displayName,
                            expiresAt: expiresAt
                        )
                    }
                }
            } catch {
                self.error = error.localizedDescription
                isLoading = false
            }
        }
    }

    func likeBack(likeId: String) {
        guard let index = likes.firstIndex(where: { $0.id == likeId }) else { return }
        let like = likes.remove(at: index)

        // Provisional match shown immediately — but with a placeholder id.
        // Opening a chat against that fake id 404s, so swap in the real
        // match_id from the action response as soon as it arrives.
        let provisionalId = "pending-\(like.user.id)"
        let expiresAt = Date().addingTimeInterval(86400)
        let match = Match(
            id: provisionalId,
            otherUser: like.user,
            status: .pendingFirstMessage,
            matchedAt: Date(),
            expiresAt: expiresAt,
            firstMsgLocked: true
        )
        matches.insert(match, at: 0)

        Task {
            if let result = try? await api.performAction(targetId: like.user.id, type: .like),
               let realId = result.matchId {
                if let idx = matches.firstIndex(where: { $0.id == provisionalId }) {
                    var fixed = matches[idx]
                    fixed = Match(
                        id: realId,
                        otherUser: fixed.otherUser,
                        status: fixed.status,
                        matchedAt: fixed.matchedAt,
                        expiresAt: fixed.expiresAt,
                        firstMsgLocked: fixed.firstMsgLocked
                    )
                    matches[idx] = fixed
                }
            } else {
                // No match id returned — fall back to the server's truth
                loadInbox()
            }
        }

        // Trigger match notification
        NotificationManager.shared.addNotification(
            type: .match,
            title: "New Match!",
            body: "You and \(like.user.displayName) matched! Say hi before the timer runs out.",
            actionData: like.user.id
        )

        // Schedule expiry reminders for this new match
        NotificationManager.shared.scheduleExpiryReminder(
            matchName: like.user.displayName,
            expiresAt: expiresAt
        )
    }

    func passLike(likeId: String) {
        guard let like = likes.first(where: { $0.id == likeId }) else { return }
        likes.removeAll { $0.id == likeId }
        // Record the pass so this liker doesn't reappear on the next refetch.
        Task { _ = try? await api.performAction(targetId: like.user.id, type: .pass) }
    }

    func unmatch(matchId: String) {
        matches.removeAll { $0.id == matchId }
    }

    /// Insert a freshly-created match immediately (e.g. from the Discover
    /// match popup) so the chat screen can open it before the next inbox
    /// refetch replaces it with the server copy.
    func upsertMatch(_ match: Match) {
        if !matches.contains(where: { $0.id == match.id }) {
            matches.insert(match, at: 0)
        }
        loadInbox()
    }

    /// Called when a reply is received after an ice breaker / first message.
    /// Moves the match from the timer-avatar section to the permanent Chats list.
    func activateMatch(id: String) {
        guard let index = matches.firstIndex(where: { $0.id == id }) else { return }
        matches[index].status = .active
        matches[index].expiresAt = nil
    }
}

File: android/app/src/main/java/com/mitimaiti/app/viewmodels/InboxViewModel.kt
package com.mitimaiti.app.viewmodels

import androidx.lifecycle.ViewModel
import androidx.lifecycle.viewModelScope
import com.mitimaiti.app.models.*
import com.mitimaiti.app.services.APIService
import com.mitimaiti.app.utils.AppNotificationManager
import com.mitimaiti.app.utils.NotificationType
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.asStateFlow
import kotlinx.coroutines.launch

class InboxViewModel : ViewModel() {
    private val _likes = MutableStateFlow<List<LikedYouCard>>(emptyList())
    val likes: StateFlow<List<LikedYouCard>> = _likes.asStateFlow()
    private val _matches = MutableStateFlow<List<Match>>(emptyList())
    val matches: StateFlow<List<Match>> = _matches.asStateFlow()
    private val _isLoading = MutableStateFlow(false)
    val isLoading: StateFlow<Boolean> = _isLoading.asStateFlow()
    private val _error = MutableStateFlow<String?>(null)
    val error: StateFlow<String?> = _error.asStateFlow()
    val totalLikes: Int get() = _likes.value.size
    val totalMatches: Int get() = _matches.value.size
    val unreadMessages: Int get() = _matches.value.sumOf { it.unreadCount }

    private var hasLoadedOnce = false

    init {
        // Real-time: the backend emits new_match / match_update over the
        // socket. Refresh the inbox and surface an in-app notification even
        // when the user isn't on the Discover screen.
        viewModelScope.launch {
            com.mitimaiti.app.services.SocketManager.shared.newMatches.collect { json ->
                val name = json.optString("displayName").ifEmpty { "Someone" }
                AppNotificationManager.shared.addNotification(
                    type = NotificationType.MATCH,
                    title = "It's a Match!",
                    body = "You and $name liked each other!"
                )
                loadInbox()
            }
        }
        viewModelScope.launch {
            com.mitimaiti.app.services.SocketManager.shared.matchUpdates.collect { json ->
                val matchId = json.optString("matchId")
                if (matchId.isNotEmpty() && json.optString("status") == "active") {
                    _matches.value = _matches.value.map { m ->
                        if (m.id == matchId) m.copy(status = MatchStatus.ACTIVE, expiresAt = null, firstMsgLocked = false)
                        else m
                    }
                }
            }
        }
    }

    fun loadInbox() {
        // Always refetch from the server (the source of truth for matches +
        // their activation state), so a new match from a Discover like appears
        // on the next inbox open. Matches iOS, which has no reload guard.
        viewModelScope.launch {
            _isLoading.value = true
            APIService.fetchInbox().onSuccess { (likes, matches) ->
                val prev = _likes.value.size
                _likes.value = likes
                _matches.value = matches
                hasLoadedOnce = true
                if (likes.size > prev && prev > 0) {
                    AppNotificationManager.shared.addNotification(
                        type = NotificationType.LIKE,
                        title = "New likes!",
                        body = "${likes.size - prev} people liked your profile"
                    )
                }
            }.onFailure { _error.value = "Failed to load inbox" }
            _isLoading.value = false
        }
    }

    fun likeBack(likeId: String) {
        val like = _likes.value.firstOrNull { it.id == likeId } ?: return; _likes.value = _likes.value.filter { it.id != likeId }
        _matches.value = listOf(Match(otherUser = like.user, status = MatchStatus.PENDING_FIRST_MESSAGE, matchedAt = System.currentTimeMillis(), expiresAt = System.currentTimeMillis() + 24 * 60 * 60 * 1000L)) + _matches.value
        AppNotificationManager.shared.addNotification(type = NotificationType.MATCH, title = "It's a Match!", body = "You and ${like.user.displayName} liked each other!")
        // Persist the like so the match is actually created server-side (they
        // already liked us → mutual match). Without this the match is cosmetic
        // and vanishes on the next inbox refetch.
        viewModelScope.launch { APIService.performAction(like.user.id, "like") }
    }

    fun passLike(likeId: String) {
        val like = _likes.value.firstOrNull { it.id == likeId } ?: return
        _likes.value = _likes.value.filter { it.id != likeId }
        // Record the pass so this liker doesn't reappear on the next refetch.
        viewModelScope.launch { APIService.performAction(like.user.id, "pass") }
    }
    fun unmatch(matchId: String) { _matches.value = _matches.value.filter { it.id != matchId } }

    /** Insert a freshly-created match immediately (e.g. from the Discover
     *  match popup) so the chat screen can open it before the next inbox
     *  refetch replaces it with the server copy. */
    fun upsertMatch(match: Match) {
        if (_matches.value.none { it.id == match.id }) {
            _matches.value = listOf(match) + _matches.value
        }
        loadInbox()
    }

    /**
     * Called when a reply is received after the ice breaker.
     * Transitions match from PENDING → ACTIVE, removes expiry (chat saved permanently).
     */
    fun activateMatch(matchId: String, lastMessage: String = "") {
        _matches.value = _matches.value.map { match ->
            if (match.id == matchId) {
                match.copy(
                    status = MatchStatus.ACTIVE,
                    expiresAt = null,
                    firstMsgLocked = false,
                    lastMessage = lastMessage.ifEmpty { match.lastMessage },
                    unreadCount = match.unreadCount + 1
                )
            } else match
        }
        val match = _matches.value.firstOrNull { it.id == matchId }
        if (match != null) {
            AppNotificationManager.shared.addNotification(
                type = NotificationType.MESSAGE,
                title = match.otherUser.displayName,
                body = lastMessage
            )
        }
    }
}
